kymatio/scattering1d/frontend/base_frontend.py

```diff
@@ -237,25 +230,68 @@ def _check_input(self, x):
                     len(x.shape)))
 
     @property
-    def J_pad(self):
-        warn("The attribute J_pad is deprecated and will be removed in v0.4. "
-        "Measure len(self.phi_f[0]) for the padded length (previously 2**J_pad) "
-        "or access shape[0] for the unpadded length (previously N).", DeprecationWarning)
-        return int(np.log2(self._N_padded))
+    def oversampling(self):
+        warn("The attribute oversampling is deprecated and will be removed in v0.5. "
+        "Pass stride = 2**(J-oversampling) or stride = 2**(log2(T)-oversampling) "
+        "to retain current behavior.", DeprecationWarning)
+        return self._oversampling
 
     @property
-    def N(self):
-        warn("The attribute N is deprecated and will be removed in v0.4. "
-        "Measure len(self.phi_f[0]) for the padded length (previously 2**J_pad) "
-        "or access shape[0] for the unpadded length (previously N).", DeprecationWarning)
-        return int(self.shape[0])
+    def average(self):
+        return parse_T(self._T, self.J, self.shape[0], T_alias='T')[1]
 
     @property
     def filterbank(self):
         filterbank_kwargs = {
             "alpha": self.alpha, "r_psi": self.r_psi, "sigma0": self.sigma0}
         return (anden_generator, filterbank_kwargs)
 
+    @property
+    def log2_stride(self):
+        if self._stride is None:
+            return self.log2_T
+        if not isinstance(self._stride, numbers.Integral):
+            raise ValueError("stride must be integer. Got: {}".format(
+                self._stride))
+        log2_stride = math.log2(self._stride)
+        if math.floor(log2_stride) != math.ceil(log2_stride):
+            raise ValueError("stride must be a power of two. Got: {}".format(
+                self._stride))
+        if self.average in [False, "global"]:
+            raise ValueError("stride={} is incompatible with T={}.".format(
+                self._stride, self._T))
+        return int(log2_stride)
+
+    @property
+    def log2_T(self):
+        return int(math.floor(math.log2(self.T)))
+
+    @property
+    def Q(self):
+        if np.any(np.array(self._Q) < 1):
+            raise ValueError('Q must always be >= 1, got {}'.format(self._Q))
+        if isinstance(self._Q, int):
+            return (self._Q, 1)
+        elif isinstance(self._Q, tuple):
+            if len(self._Q) == 1:
+                return (self._Q + (1,))
+            elif len(self._Q) == 2:
+                return self._Q
+            raise NotImplementedError("Q must be an integer, 1-tuple or "
```

Comment:
can you add a test that raises this error?

Reply:
on it

Reply:
90cbb6b

Reply:
90cbb6b